xain_fl/coordinator/coordinator.py

```diff
@@ -236,6 +238,17 @@ def select_participant_ids_and_init_round(self) -> None:
         selected_ids = self.controller.select_ids(self.participants.ids())
         self.round = Round(selected_ids)
 
+    def select_outstanding(self) -> List[str]:
+        """Selects participant ids."""
+
+        selected = set(self.round.participant_ids)
+        num_outstanding = self.minimum_participants_in_round - len(selected)
+        pool = set(self.participants.ids()) - selected
+        frac = num_outstanding / len(pool)
+
+        self.controller.fraction_of_participants = frac
```

Comment:
Is it ok to override this? Isn't it a value that comes from the coordinator config file.

Reply:
I also don't see `fraction_of_participants` being used at all here. 
Say we have `fraction_of_participants = 0.5` and `minimum_participants_in_round = 10`.
Here is a scenario:

1. 20 participants connect => 10 are selected, and round start
2. 1 of the selected participant drops out, along with 9 of the non-selected participants => the selected participant is removed from the round and the participants list, and the 9 non-selected participants are removed from the participants list. Finally `select_outstanding()` is called
3. we have `selected = 9`, `num_outstanding = 1` and `pool = 1`, so `frac = 1` and we'll select 1 participant our of a pool of 1.

My question: shouldn't we only proceed to a new selection if we have enough connected participants such that `num_outstanding / pool <= fraction_of_participants`?

Reply:
yes, this is a bit unfortunate as the controller "API" isn't currently so great. It would have been easier if we could "select `n` from `m`" rather than insisting on setting a fraction. But I didn't want to tackle the controller API here, though it could be a separate issue.

Reply:
Sounds good. What about my second question: is it ok when re-selecting new participants not to honor `num_outstanding / pool <= fraction_of_participants` ?

Reply:
actually, we must ignore this check. the reason is that both coordinator and controller have a `fraction_of_participants` attribute, but they refer to two different things. the coordinator uses `fraction_of_participants` as a ratio for all participants, but the controller uses this as a ratio for the participants relative to the respective set (which can for example be all participants, but also only outstanding participants).

i fully agree that we have to reorganize the controller api, as it's current usage is pretty obscure. but ofcourse this should be done in a dedicated merge request.

Reply:
sorry I hadn't refreshed, missed the 2nd question.

In your scenario, 20 connect and 10 are selected. One of those selected drops (along with a bunch of the others), so the round is paused. Currently, the round resumes only when we have `minimum_connected_participants` again i.e. 20. At this point, we will be selecting 1 outstanding out of a pool of 11.

More generally, because we only re-select when `minimum_connected_participants`, even in the worst case (where we must re-select everyone) we're just back to `fraction_of_participants`.

Reply:
Ah ok thanks for the explanation!

Reply:
@janpetschexain I created a ticket for the controller API issue https://xainag.atlassian.net/browse/PB-433 cheers